--- Day 21: Keypad Conundrum ---
As you teleport onto Santa's Reindeer-class starship, The Historians begin to panic: someone from their search party is missing. A quick life-form scan by the ship's computer reveals that when the missing Historian teleported, he arrived in another part of the ship.

The door to that area is locked, but the computer can't open it; it can only be opened by physically typing the door codes (your puzzle input) on the numeric keypad on the door.

The numeric keypad has four rows of buttons: 789, 456, 123, and finally an empty gap followed by 0A. Visually, they are arranged like this:

+---+---+---+
| 7 | 8 | 9 |
+---+---+---+
| 4 | 5 | 6 |
+---+---+---+
| 1 | 2 | 3 |
+---+---+---+
    | 0 | A |
    +---+---+
Unfortunately, the area outside the door is currently depressurized and nobody can go near the door. A robot needs to be sent instead.

The robot has no problem navigating the ship and finding the numeric keypad, but it's not designed for button pushing: it can't be told to push a specific button directly. Instead, it has a robotic arm that can be controlled remotely via a directional keypad.

The directional keypad has two rows of buttons: a gap / ^ (up) / A (activate) on the first row and < (left) / v (down) / > (right) on the second row. Visually, they are arranged like this:

    +---+---+
    | ^ | A |
+---+---+---+
| < | v | > |
+---+---+---+
When the robot arrives at the numeric keypad, its robotic arm is pointed at the A button in the bottom right corner. After that, this directional keypad remote control must be used to maneuver the robotic arm: the up / down / left / right buttons cause it to move its arm one button in that direction, and the A button causes the robot to briefly move forward, pressing the button being aimed at by the robotic arm.

For example, to make the robot type 029A on the numeric keypad, one sequence of inputs on the directional keypad you could use is:

< to move the arm from A (its initial position) to 0.
A to push the 0 button.
^A to move the arm to the 2 button and push it.
>^^A to move the arm to the 9 button and push it.
vvvA to move the arm to the A button and push it.
In total, there are three shortest possible sequences of button presses on this directional keypad that would cause the robot to type 029A: <A^A>^^AvvvA, <A^A^>^AvvvA, and <A^A^^>AvvvA.

Unfortunately, the area containing this directional keypad remote control is currently experiencing high levels of radiation and nobody can go near it. A robot needs to be sent instead.

When the robot arrives at the directional keypad, its robot arm is pointed at the A button in the upper right corner. After that, a second, different directional keypad remote control is used to control this robot (in the same way as the first robot, except that this one is typing on a directional keypad instead of a numeric keypad).

There are multiple shortest possible sequences of directional keypad button presses that would cause this robot to tell the first robot to type 029A on the door. One such sequence is v<<A>>^A<A>AvA<^AA>A<vAAA>^A.

Unfortunately, the area containing this second directional keypad remote control is currently -40 degrees! Another robot will need to be sent to type on that directional keypad, too.

There are many shortest possible sequences of directional keypad button presses that would cause this robot to tell the second robot to tell the first robot to eventually type 029A on the door. One such sequence is <vA<AA>>^AvAA<^A>A<v<A>>^AvA^A<vA>^A<v<A>^A>AAvA^A<v<A>A>^AAAvA<^A>A.

Unfortunately, the area containing this third directional keypad remote control is currently full of Historians, so no robots can find a clear path there. Instead, you will have to type this sequence yourself.

Were you to choose this sequence of button presses, here are all of the buttons that would be pressed on your directional keypad, the two robots' directional keypads, and the numeric keypad:

<vA<AA>>^AvAA<^A>A<v<A>>^AvA^A<vA>^A<v<A>^A>AAvA^A<v<A>A>^AAAvA<^A>A
v<<A>>^A<A>AvA<^AA>A<vAAA>^A
<A^A>^^AvvvA
029A
In summary, there are the following keypads:

One directional keypad that you are using.
Two directional keypads that robots are using.
One numeric keypad (on a door) that a robot is using.
It is important to remember that these robots are not designed for button pushing. In particular, if a robot arm is ever aimed at a gap where no button is present on the keypad, even for an instant, the robot will panic unrecoverably. So, don't do that. All robots will initially aim at the keypad's A key, wherever it is.

To unlock the door, five codes will need to be typed on its numeric keypad. For example:

029A
980A
179A
456A
379A
For each of these, here is a shortest sequence of button presses you could type to cause the desired code to be typed on the numeric keypad:

029A: <vA<AA>>^AvAA<^A>A<v<A>>^AvA^A<vA>^A<v<A>^A>AAvA^A<v<A>A>^AAAvA<^A>A
980A: <v<A>>^AAAvA^A<vA<AA>>^AvAA<^A>A<v<A>A>^AAAvA<^A>A<vA>^A<A>A
179A: <v<A>>^A<vA<A>>^AAvAA<^A>A<v<A>>^AAvA^A<vA>^AA<A>A<v<A>A>^AAAvA<^A>A
456A: <v<A>>^AA<vA<A>>^AAvAA<^A>A<vA>^A<A>A<vA>^A<A>A<v<A>A>^AAvA<^A>A
379A: <v<A>>^AvA^A<vA<AA>>^AAvA<^A>AAvA^A<vA>^AA<A>A<v<A>A>^AAAvA<^A>A
The Historians are getting nervous; the ship computer doesn't remember whether the missing Historian is trapped in the area containing a giant electromagnet or molten lava. You'll need to make sure that for each of the five codes, you find the shortest sequence of button presses necessary.

The complexity of a single code (like 029A) is equal to the result of multiplying these two values:

The length of the shortest sequence of button presses you need to type on your directional keypad in order to cause the code to be typed on the numeric keypad; for 029A, this would be 68.
The numeric part of the code (ignoring leading zeroes); for 029A, this would be 29.
In the above example, complexity of the five codes can be found by calculating 68 * 29, 60 * 980, 68 * 179, 64 * 456, and 64 * 379. Adding these together produces 126384.

Find the fewest number of button presses you'll need to perform in order to cause the robot in front of the door to type each code. What is the sum of the complexities of the five codes on your list?

In [ ]:
"""
Initial Notes:
<vA<AA>>^AvAA<^A>A<v<A>>^AvA^A<vA>^A<v<A>^A>AAvA^A<v<A>A>^AAAvA<^A>A
  v <<   A >>  ^ A   <   A > A  v  A   <  ^ AA > A   < v  AAA >  ^ A
         <       A       ^   A     >        ^^   A        vvv      A
                 0           2                   9                 A


What I am thinking about is: What makes a path optimal? What are the constraints?

On the highest level (numpad) the sequence is predetermined. 
On each subsequent level, given the required input sequence for the higher level, we know exactly which inputs we need, but there can be some variability regarding the order of inputs (eg do i go v< or <v to move from 9 to 5)

This variability can compound over multiple level, such as the difference between Av<A and A<vA

However, each "A" provides a natural reset where all lower level robots will need to press "A" as well. Thus we can chunk sequences where all robots start on A and end on A. We can compute costs for simple sequences and chain those to compute more elaborate, multi-level sequences.

In the end we only need to compute one optimal sequence for each num key sequence (start on button X, end with pressing button Y) since all higher level keypads start and end on A

Im gonna assume that an ordered sequence (eg vv>>) will always be weakly better than unordered (eg v>v>)
"""

'\n<vA<AA>>^AvAA<^A>A<v<A>>^AvA^A<vA>^A<v<A>^A>AAvA^A<v<A>A>^AAAvA<^A>A\n  v <<   A >>  ^ A   <   A > A  v  A   <  ^ AA > A   < v  AAA >  ^ A\n         <       A       ^   A     >        ^^   A        vvv      A\n                 0           2                   9                 A\n\n\nWhat I am thinking about is: What makes a path optimal? What are the constraints?\n\nOn the highest level (numpad) the sequence is predetermined. \nOn each subsequent level, given the required input sequence for the higher level, we know exactly which inputs we need, but there can be some variability regarding the order of inputs (eg do i go v< or <v to move from 9 to 5)\n\nThis variability can compound over multiple level, such as the difference between Av<A and A<vA\n\nHowever, each "A" provides a natural reset where all lower level robots will need to press "A" as well. Thus we can chunk sequences where all robots start on A and end on A. We can compute costs for simple sequences and chain those to co

In [2]:
def coord_to_seq(start, end, gap=None):
    s_x, s_y = start
    e_x, e_y = end

    hor = vert = ""

    if e_x > s_x:
        vert = "v" * abs(e_x - s_x)
    elif s_x > e_x:
        vert = "^" * abs(e_x - s_x)
    if e_y > s_y:
        hor = ">" * abs(s_y - e_y)
    elif e_y < s_y:
        hor = "<" * abs(s_y - e_y)

    hv = hor + vert
    vh = vert + hor

    if gap:
        g_x, g_y = gap
        if s_x == g_x and e_y == g_y:
            return [vh]
        elif s_y == g_y and e_x == g_x:
            return [hv]

    if hv != vh:
        return [vh, hv]
    else:
        return [vh]
coord_to_seq((0,0), (3,5), (3,0))

['>>>>>vvv']

In [3]:
def num_to_key(start, end):
    coord = {"7" : (0,0),
             "8" : (0,1),
             "9" : (0,2),
             "4" : (1,0),
             "5" : (1,1),
             "6" : (1,2),
             "1" : (2,0),
             "2" : (2,1),
             '3' : (2,2),
             "gap" : (3,0),
             '0'   : (3,1),
             "A"   : (3,2),}
    
    s = coord[start]
    e = coord[end]
    gap = coord["gap"]
    paths = coord_to_seq(s, e, gap)
    return [path + "A" for path in paths]

def key_to_key(start,end):
    coord = {"gap" : (0,0),
             "^" : (0,1),
             "A" : (0,2),
             "<" : (1,0),
             "v" : (1,1),
             ">" : (1,2),
             }
    s = coord[start]
    e = coord[end]
    gap = coord["gap"]
    paths = coord_to_seq(s, e, gap)
    return [path + "A" for path in paths]

num_to_key("5","9"), key_to_key("A", "v")

(['^>A', '>^A'], ['v<A', '<vA'])

In [4]:
from itertools import product

def ktk_exh(seq):
    path_list = []
    for start, end in zip("A" + seq, seq):
        path_list.append(key_to_key(start, end))
    return ["".join(p) for p in product(*path_list)]

def calc_opt(start, end):
    key1 = num_to_key(start, end)
    key2 = [ktk_exh(x) for x in key1]
    key2 = [string for sublist in key2 for string in sublist]
    key3 = [ktk_exh(x) for x in key2]
    key3 = [string for sublist in key3 for string in sublist]
    return min(key3, key=len)


In [5]:
num_buttons = "0123456789A"
opt_seq = {}
for start in num_buttons:
    for end in num_buttons:
        opt_seq[start + end] = calc_opt(start, end)
opt_seq


{'00': 'A',
 '01': 'v<<A>>^Av<A<A>>^AvAA^<A>A',
 '02': 'v<<A>>^AvA^A',
 '03': 'v<<A>>^Av<A>A^A<A>A',
 '04': 'v<<A>>^AAv<A<A>>^AvAA^<A>A',
 '05': 'v<<A>>^AAvA^A',
 '06': 'v<<A>>^AAv<A>A^A<A>A',
 '07': 'v<<A>>^AAAv<A<A>>^AvAA^<A>A',
 '08': 'v<<A>>^AAAvA^A',
 '09': 'v<<A>>^AAAv<A>A^A<A>A',
 '0A': 'v<A^>A<A>A',
 '10': 'v<A^>Av<<A>>^A<Av>A^A',
 '11': 'A',
 '12': 'v<A^>A<A>A',
 '13': 'v<A^>AA<A>A',
 '14': 'v<<A>>^AvA^A',
 '15': 'v<<A>>^Av<A>A^A<A>A',
 '16': 'v<<A>>^Av<A>A^AA<A>A',
 '17': 'v<<A>>^AAvA^A',
 '18': 'v<<A>>^AAv<A>A^A<A>A',
 '19': 'v<<A>>^AAv<A>A^AA<A>A',
 '1A': 'v<A^>AAv<<A>>^A<Av>A^A',
 '20': 'v<A<A>>^A<Av>A^A',
 '21': 'v<A<AA>>^AvAA^<A>A',
 '22': 'A',
 '23': 'v<A^>A<A>A',
 '24': 'v<A<AA>>^AvA^<A>AvA^A',
 '25': 'v<<A>>^AvA^A',
 '26': 'v<<A>>^Av<A>A^A<A>A',
 '27': 'v<A<AA>>^AvA^<A>AAvA^A',
 '28': 'v<<A>>^AAvA^A',
 '29': 'v<<A>>^AAv<A>A^A<A>A',
 '2A': 'v<A<A>>^AvA^A<A>A',
 '30': 'v<A<AA>>^AvA^A<Av>A^A',
 '31': 'v<A<AA>>^AAvAA^<A>A',
 '32': 'v<A<AA>>^AvAA^<A>A',
 '33': 'A',
 '34': 

In [6]:
code = '980A'
btn_presses = ""
for pair in zip("A" + code, code):
    pair = "".join(pair)
    btn_presses += opt_seq[pair]
btn_presses, len(btn_presses)

('v<<A>>^AAAvA^Av<A<AA>>^AvAA^<A>Av<A<A>>^AAA<Av>A^Av<A^>A<A>A', 60)

In [7]:
puzzle_input = """973A
836A
780A
985A
413A"""
codes = puzzle_input.splitlines()
codes

['973A', '836A', '780A', '985A', '413A']

In [8]:
sum_compl = 0
for code in codes:
    btn_presses = ""
    for pair in zip("A" + code, code):
        pair = "".join(pair)
        btn_presses += opt_seq[pair]
    num_code = int(code[:3])
    compl = num_code * len(btn_presses)
    sum_compl += compl
sum_compl

270084

Your puzzle answer was 270084.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
Just as the missing Historian is released, The Historians realize that a second member of their search party has also been missing this entire time!

A quick life-form scan reveals the Historian is also trapped in a locked area of the ship. Due to a variety of hazards, robots are once again dispatched, forming another chain of remote control keypads managing robotic-arm-wielding robots.

This time, many more robots are involved. In summary, there are the following keypads:

One directional keypad that you are using.
25 directional keypads that robots are using.
One numeric keypad (on a door) that a robot is using.
The keypads form a chain, just like before: your directional keypad controls a robot which is typing on a directional keypad which controls a robot which is typing on a directional keypad... and so on, ending with the robot which is typing on the numeric keypad.

The door codes are the same this time around; only the number of robots and directional keypads has changed.

Find the fewest number of button presses you'll need to perform in order to cause the robot in front of the door to type each code. What is the sum of the complexities of the five codes on your list?

In [9]:
"""
We will need an efficient and robust algorithm for this part. 

Let's concentrate on the keypad to keypad conversion first. I want to create a dictionary for the optimal one level expansion for every sequence (start position) - (end position) - (press A). Eg if the arm starts at "A", needs to move to "v" and then activate the expansion sequence is "<vA" or "v<A". Since every such sequence ends on "A", the optimal expansion does not depend on outside context - the only thing we need is to specify start and end. 

For many sequences, there is really only one possible expansion, eg start(<), end(v) only ever expands into >A. For those with multiple values, we will have to figure out what's going on to ensure our dict will have the optimal one. For instance, arranging the sequence on the next or subsequent levels such that buttons farther from A are pressed in succession provides speedups. 

Once we have this dictionary, we can simply count the expansions, eg if we have one A -> v -> press passage in our full sequence, we can use the dict to simply count how many other x->y->press couplings we will have at the next level and so on.
"""

'\nWe will need an efficient and robust algorithm for this part. \n\nLet\'s concentrate on the keypad to keypad conversion first. I want to create a dictionary for the optimal one level expansion for every sequence (start position) - (end position) - (press A). Eg if the arm starts at "A", needs to move to "v" and then activate the expansion sequence is "<vA" or "v<A". Since every such sequence ends on "A", the optimal expansion does not depend on outside context - the only thing we need is to specify start and end. \n\nFor many sequences, there is really only one possible expansion, eg start(<), end(v) only ever expands into >A. For those with multiple values, we will have to figure out what\'s going on to ensure our dict will have the optimal one. For instance, arranging the sequence on the next or subsequent levels such that buttons farther from A are pressed in succession provides speedups. \n\nOnce we have this dictionary, we can simply count the expansions, eg if we have one A ->

In [24]:
ktk_expansions = {}
for start in "<>v^A":
    for end in "<>v^A":
        expansions = key_to_key(start, end)
        if len(expansions) == 1:
            ktk_expansions[start+end] = expansions[0]
        else:
            print(f"{start}{end}: {expansions}")

>^: ['^<A', '<^A']
vA: ['^>A', '>^A']
^>: ['v>A', '>vA']
Av: ['v<A', '<vA']


In [ ]:
# All of these kinda come down to: If I have to go horizontal and vertical, which direction should i go first?
# On the next level, we want to go left before going up/down, since this means we can move << from A
ktk_expansions[">^"] = '<^A'
ktk_expansions["Av"] = '<vA'

# This leaves ^>: ['v>A', '>vA'] as well as vA: ['^>A', '>^A']
# With ^>, the expansion v>A is preferable. Why? In order to visit v, > and A we have to move left, down, right and up exactly once (left + down to reach v, before going back up to A. We hit up > somewhere in between). As said earlier, if we have to go left, it is good to do it immediately (coming from A), since this allows us to double tap << on the expansion.
ktk_expansions["^>"] = 'v>A'

# (Note: I did some more exploration to find the correct order for the last three, but omitted that here to make the notebook a liiitle bit leaner)
# The last holdous is vA: ['^>A', '>^A']


In [12]:
# Let's write code that finds optimal expansions for the last case (expanding vA)

def expand_seq(seq, exp_vA, init="A"):
    expansion = ""
    for pair in zip(init + seq, seq):
        pair = "".join(pair)
        if pair == 'vA':
            expansion += exp_vA
        else:
            expansion += ktk_expansions[pair]
    return expansion

In [13]:
exp_va1 = '^>A'
exp_va2 = '>^A'
seq = 'vA'
for _ in range(5):
    seq = expand_seq(seq, exp_va1)
    print(len(seq), seq)

6 <vA^>A
16 v<<A>A^>A<Av>A^A
40 <vA<AA>>^AvA^A<Av>A^Av<<A>>^A<vA>A^A<A>A
102 v<<A>A^>Av<<A>>^AAvAA<^A>A<vA^>A<A>Av<<A>>^A<vA>A^A<A>A<vA<AA>>^AvAA<^A>Av<<A>A^>AvA^A<A>Av<<A>>^AvA^A
252 <vA<AA>>^AvA^A<Av>A^A<vA<AA>>^AvAA<^A>AA<vA^>AAv<<A>^A>AvA^Av<<A>A^>A<Av>A^Av<<A>>^AvA^A<vA<AA>>^AvAA<^A>Av<<A>A^>AvA^A<A>Av<<A>>^AvA^Av<<A>A^>Av<<A>>^AAvAA<^A>A<vA^>AAv<<A>^A>AvA^A<vA<AA>>^AvA^A<Av>A^A<vA^>A<A>Av<<A>>^AvA^A<vA<AA>>^AvAA<^A>A<vA^>A<A>A


In [14]:
seq = 'vA'
for _ in range(5):
    seq = expand_seq(seq, exp_va2)
    print(len(seq), seq)

6 <vA>^A
16 v<<A>A>^AvA<^A>A
40 <vA<AA>>^AvA^AvA<^A>A<vA>^Av<<A>^A>AvA^A
102 v<<A>A>^Av<<A>>^AAvAA<^A>A<vA>^A<A>A<vA>^Av<<A>^A>AvA^Av<<A>A>^AvA<^A>A<vA<AA>>^AvA<^A>AvA^A<vA>^A<A>A
258 <vA<AA>>^AvA^AvA<^A>A<vA<AA>>^AvAA<^A>AA<vA>^AAv<<A>^A>AvA^Av<<A>A>^AvA<^A>Av<<A>>^AvA^Av<<A>A>^AvA<^A>A<vA<AA>>^AvA<^A>AvA^A<vA>^A<A>A<vA<AA>>^AvA^AvA<^A>A<vA>^Av<<A>^A>AvA^Av<<A>A>^Av<<A>>^AAvAA<^A>A<vA>^Av<<A>^A>AvA^A<vA>^A<A>Av<<A>A>^AvA<^A>Av<<A>>^AvA^A


In [64]:
seq = "^>"
for _ in range(5):
    seq = expand_seq(seq)
    print(len(seq), seq)

5 <Av>A
15 v<<A>>^A<vA>A^A
35 <vA<AA>>^AvAA<^A>Av<<A>A^>AvA^A<A>A
89 v<<A>A^>Av<<A>>^AAvAA<^A>A<vA^>AAv<<A>^A>AvA^A<vA<AA>>^AvA^A<Av>A^A<vA^>A<A>Av<<A>>^AvA^A
219 <vA<AA>>^AvA^A<Av>A^A<vA<AA>>^AvAA<^A>AA<vA^>AAv<<A>^A>AvA^Av<<A>A^>A<Av>A^AA<vA<AA>>^AvA<^A>AvA^A<vA^>A<A>Av<<A>A^>Av<<A>>^AAvAA<^A>A<vA^>A<A>Av<<A>>^A<vA>A^A<A>Av<<A>A^>A<Av>A^Av<<A>>^AvA^A<vA<AA>>^AvAA<^A>A<vA^>A<A>A


In [65]:
seq = ">^"
for _ in range(5):
    seq = expand_seq(seq)
    print(len(seq), seq)

5 vA<^A
15 <vA^>Av<<A>^A>A
37 v<<A>A^>A<Av>A^A<vA<AA>>^AvA<^A>AvA^A
95 <vA<AA>>^AvA^A<Av>A^Av<<A>>^A<vA>A^A<A>Av<<A>A^>Av<<A>>^AAvAA<^A>A<vA^>Av<<A>^A>AvA^A<vA^>A<A>A
237 v<<A>A^>Av<<A>>^AAvAA<^A>A<vA^>A<A>Av<<A>>^A<vA>A^A<A>A<vA<AA>>^AvAA<^A>Av<<A>A^>AvA^A<A>Av<<A>>^AvA^A<vA<AA>>^AvA^A<Av>A^A<vA<AA>>^AvAA<^A>AA<vA^>AAv<<A>^A>AvA^Av<<A>A^>A<Av>A^A<vA<AA>>^AvA<^A>AvA^A<vA^>A<A>Av<<A>A^>A<Av>A^Av<<A>>^AvA^A


In [ ]:
# ok so the first expansion wins after quite a number of expansions. I am not gonna analyze the reason for this deeply but just gonna roll with it
ktk_expansions['vA'] = '^>A'

In [19]:
# let's clean up the expand function, we're gonna need it:
def expand_seq(seq, init="A"):
    expansion = ""
    for pair in zip(init + seq, seq):
        pair = "".join(pair)
        expansion += ktk_expansions[pair]
    return expansion

In [23]:
# Now that we finally have a complete key_to_key dictionary...let's make our num to key dictionary
ntk_expansions = {}
for start in "0123456789A":
    for end in "0123456789A":
        expansions = num_to_key(start, end)
        if len(expansions) == 1:
            ntk_expansions[start+end] = expansions[0]
        else:
            undecided = True
            mutated = expansions.copy()
            while undecided:
                mutated = [expand_seq(m) for m in mutated]
                m_len = [len(seq) for seq in mutated]
                if m_len.count(min(m_len)) == 1:
                    undecided = False
                    min_idx = m_len.index(min(m_len))
            ntk_expansions[start+end] = expansions[min_idx]
print(ntk_expansions)

{'00': 'A', '01': '^<A', '02': '^A', '03': '^>A', '04': '^^<A', '05': '^^A', '06': '^^>A', '07': '^^^<A', '08': '^^^A', '09': '^^^>A', '0A': '>A', '10': '>vA', '11': 'A', '12': '>A', '13': '>>A', '14': '^A', '15': '^>A', '16': '^>>A', '17': '^^A', '18': '^^>A', '19': '^^>>A', '1A': '>>vA', '20': 'vA', '21': '<A', '22': 'A', '23': '>A', '24': '<^A', '25': '^A', '26': '^>A', '27': '<^^A', '28': '^^A', '29': '^^>A', '2A': 'v>A', '30': '<vA', '31': '<<A', '32': '<A', '33': 'A', '34': '<<^A', '35': '<^A', '36': '^A', '37': '<<^^A', '38': '<^^A', '39': '^^A', '3A': 'vA', '40': '>vvA', '41': 'vA', '42': 'v>A', '43': 'v>>A', '44': 'A', '45': '>A', '46': '>>A', '47': '^A', '48': '^>A', '49': '^>>A', '4A': '>>vvA', '50': 'vvA', '51': '<vA', '52': 'vA', '53': 'v>A', '54': '<A', '55': 'A', '56': '>A', '57': '<^A', '58': '^A', '59': '^>A', '5A': 'vv>A', '60': '<vvA', '61': '<<vA', '62': '<vA', '63': 'vA', '64': '<<A', '65': '<A', '66': 'A', '67': '<<^A', '68': '<^A', '69': '^A', '6A': 'vvA', '70': 

In [ ]:
# Now we are ready for the final push:
# First we prep our ktk_mappings dict from ktk_expansions for slightly easier handling later:
ktk_mappings = {}
for seed, expa in ktk_expansions.items():
    mappings = ["".join(pair) for pair in zip("A"+expa, expa)]
    ktk_mappings[seed] = mappings

ntk_mappings = {}
for seed, expa in ntk_expansions.items():
    mappings = ["".join(pair) for pair in zip("A"+expa, expa)]
    ntk_mappings[seed] = mappings

ktk_expansions['^<'], ktk_mappings['^<'], ntk_expansions["04"], ntk_mappings["04"]

('v<A', ['Av', 'v<', '<A'], '^^<A', ['A^', '^^', '^<', '<A'])

In [43]:
def num_press(code, num_keypads):
    seq_count = dict.fromkeys(ktk_expansions, 0)

    # get initial seq_counts going (num to keypad conversion):
    for pair in zip("A" + code, code):
        pair = "".join(pair)
        for keypad_pair in ntk_mappings[pair]:
            seq_count[keypad_pair] += 1

    # expand through all the keypad conversions:
    for _ in range(num_keypads - 1):
        new_count = dict.fromkeys(ktk_expansions, 0)
        for old_pair, count in seq_count.items():
            for exp_pair in ktk_mappings[old_pair]:
                new_count[exp_pair] += count
        seq_count = new_count
    return sum(seq_count.values())

num_press("029A", 3)

68

In [63]:
# main loop:
num_keypads = 26
sum_compl = 0
for code in codes:
    num_code = int(code[:3])
    compl = num_code * num_press(code, num_keypads)
    sum_compl += compl
print(f'{sum_compl:e}, {sum_compl}')

3.294310e+14, 329431019997766


That's the right answer! You are one gold star closer to finding the Chief Historian.

You have completed Day 21! You can [Share] this victory or [Return to Your Advent Calendar].